Wabt simd v128.const instruction initial PR:

lizhengxing opened the pull request:
This is an initial PR for Wabt simd support to get some early feedback.
   This PR only implement the v128.const instruction.

   It will construct a simd 128-bits const.
   The expected wat format is:
   v128.const i32 0x00000000 0x11111111 0x22222222 0x33333333

   For simplify, this PR only implement the wat2wasm and wasm2wat functions.

   The following PRs will implement the full functions in Wabt.
   such as: Interp, objdump, logging, etc....

lizhengxing commented:
@binji Review please!
@arunetm fyi

binji commented on src/wast-parser.cc:
```diff
@@ -1479,6 +1480,14 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
     }
 
+    // Simd
+    case TokenType::SimdConst: {
```
I think it'd be better to return a TokenType::Const but with a different opcode, like the other const instructions.

lizhengxing replied:
Yes.  I do wanted to implement v128.const like this way at beginning.

But parse the simd const lists is slightly different to Int32/int64/F32/F64 const.  I'm worried that the changes will mess up the original code.  That's why I use a dedicated token for simd const.

Anyway, I'm re-implementing it and try to keep the code as clean as current.

binji replied:
Yeah, I think you'll have to restructure `ParseConst` a bit, but that's OK. Don't worry too much about messing up the code. :-)

binji commented on src/wast-parser.cc:
```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+// Simd
+// The expected V128 const lists is:
+// i32 0xXXXXXXXX 0xXXXXXXXX 0xXXXXXXXX 0xXXXXXXXX
+Result WastParser::ParseSimdConstList(ConstVector* consts) {
```
Probably better if this had a different name; currently "ParseConstList" name is used to parse a list of constant values for assertions, like `(i32.const 1) (f32.const 2)` etc.

This only seems to be used by `ParseSimdConst` below, maybe it can just be inlined? Or do you think you'll need it elsewhere?

lizhengxing replied:
Inline it will cause ParseSimdConst Function bigger.  Let me try it.

binji commented on src/wast-parser.cc:
```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+    }
+
+    default:
+      assert(!"ParseSimdConstListcalled with invalid input const type");
```
This should be an `Error`, not an assert.

binji commented on src/wast-parser.cc:
```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+      // 2. Copy it into const_.v128_bits.
+      union i128bits{
+        int32_t i32Array[4];
+        uint8_t i8Array[16];
```
Why is the union needed? I think you can use the i32array directly in memcpy below.

binji commented on src/wast-parser.cc:
```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+      assert(nSize == 4);
+      // 2. Copy it into const_.v128_bits.
+      union i128bits{
+        int32_t i32Array[4];
```
Should be uint32_t, I think

binji commented on src/wast-parser.cc:
```diff
@@ -1488,6 +1497,96 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+  return Result::Ok;
+}
+
+Result WastParser::ParseSimdConst(Const* const_) {
```
I think it would be better if this was called from `ParseConst` so we could share places where that is used currently.

binji commented on src/wat-writer.cc:
```diff
@@ -452,6 +452,19 @@ void WatWriter::WriteConst(const Const& const_) {
+      int32_t i32Array[4];
+      memcpy(&i32Array, &(const_.v128_bits), sizeof(v128));
+      WritePutsSpace(Opcode::V128Const_Opcode.GetName());
+      Writef("i32 0x%08x", i32Array[0]);
```
You may as well roll this into one `Writef` call, I think

lizhengxing commented on src/common.h:
```diff
@@ -90,6 +90,9 @@
 #define PRIoffset PRIzx
 
+// Simd.
+typedef __m128 v128;
```
I'm not sure what's "define this internally" means.  Move "typedef __m128 v128" to other header file?

BTW:
How should we deal with the platform that doesn't support 128-bit integer?

binji replied:
Sorry, I meant define something in wabt directly, like:
```
struct v128 {
  uint32_t v[0];
};
```

We don't really need to use the native type, since most of the time we'll just be loading and storing the values anyway. The interpreter may want to use a real v128 type, but even there it might be nice to do it with scalar values instead.

lizhengxing commented:
@binji 
Thanks for your review and nice comments. please check my questions about the comments. thanks again.

@arunetm FYI

lizhengxing commented:
@binji 
Sorry, Some other issues blocked me submit the updated PR later.
Please review it if you are available, thanks!

The main changes are:
1. updated the PR according to upstream's comments.
2. added some tests.

@arunetm FYI

binji commented on src/common.h:
```diff
@@ -90,6 +90,10 @@
 #define PRIaddress "u"
 #define PRIoffset PRIzx
 
+typedef struct v128_data {
```
This is C++, so you can just do `struct v128 { ... };`

lizhengxing replied:
Agree.

binji commented on src/wat-writer.cc:
```diff
@@ -485,6 +485,15 @@ void WatWriter::WriteConst(const Const& const_) {
 
+    case Type::V128: {
+      int32_t i32Array[4];
+      memcpy(&i32Array, &(const_.v128_bits), sizeof(v128));
```
Remove and just use `const_.v128_bits.v[0]`, etc. below.

lizhengxing replied:
Agree.

binji commented on src/wast-parser.cc:
```diff
@@ -1499,28 +1499,85 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+    }
+
+    default:
+      Error(const_->loc, "ParseSimdConst called with invalid input const type, Type::I32 expected");
```
Since this will be displayed to the user, it's better to avoid internal function names. Maybe just "expected i32 at start of simd constant". What do you think?

lizhengxing replied:
Agree.

binji commented on src/wast-parser.cc:
```diff
@@ -1545,13 +1602,23 @@ Result WastParser::ParseConst(Const* const_) {
+    case Opcode::V128Const:
+      const_->type = Type::V128;
+      // Parse V128 Simd Const (16 bytes).
+      result = ParseSimdConst(const_, in_type, sizeof(v128));
```
Since this is always going to be `size(v128)` for now, probably best to just hard code it in `ParseSimdConst`.

lizhengxing replied:
The reason here is for future possible 256, 512 simd type too.

binji commented on src/wast-parser.cc:
```diff
@@ -1499,28 +1499,85 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+  switch (in_type) {
+    case Type::I32: {
+      const_->loc = GetLocation();
+      int nCount = nSimdConstBytes/sizeof(uint32_t);
```
We can hardcode this for now, since we know that `nSimdConstBytes` is always `sizeof(v128)`.

lizhengxing replied:
The reason why I didn't use hardcode here is: 
I want to this function can be used for future 256 or 512 simd type.

binji commented on src/wast-parser.cc:
```diff
@@ -1499,28 +1499,85 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+  switch (in_type) {
+    case Type::I32: {
+      const_->loc = GetLocation();
+      int nCount = nSimdConstBytes/sizeof(uint32_t);
```
nit: `count` instead of `nCount`

lizhengxing replied:
Agree.

binji commented on src/binary-reader.cc:
```diff
@@ -391,6 +397,14 @@ Result BinaryReader::ReadInitExpr(Index index, bool require_i32) {
     }
 
+    case Opcode::V128Const: {
+      v128 value_bits;
```
Since SIMD is a future feature, we want to only allow it if the feature flag is enabled. See other examples using the `ERROR_UNLESS_OPCODE_ENABLED` macro in this file.

lizhengxing replied:
Agree.

binji commented on src/binary-reader.cc:
```diff
@@ -629,6 +643,15 @@ Result BinaryReader::ReadFunctionBody(Offset end_offset) {
       }
 
+      case Opcode::V128Const: {
+        v128 value_bits;
```
Same here.

lizhengxing replied:
Agree.

binji commented on src/wast-parser.cc:
```diff
@@ -1545,13 +1602,23 @@ Result WastParser::ParseConst(Const* const_) {
       break;
 
+    case Opcode::V128Const:
+      const_->type = Type::V128;
```
We only want to allow parsing this if the SIMD feature is enabled, see other examples in this file using `ErrorUnlessOpcodeEnabled`.

lizhengxing replied:
Agree.

lizhengxing commented:
@binji 
Thanks for your review. I will update the PR. thanks!

@arunetm FYI!

lizhengxing commented:
@binji 
I updated the PR according to your suggestion except for the hardcode for simd value size.

BTW:
1. Running test/run-tests.py --enable-simd -v --bindir out/clang/Debug/ will meet 
"run-tests.py: error: unrecognized arguments: --enable-simd"

Would you please take a look for it?

2. Some tests may be mixed simd and non-simd functions, How should I control the --enable-simd test option when I submit a PR.

Thanks!

@arunetm FYI!

binji commented:
You can specify the flags in the test file itself, so in `test/parse/expr/bad-const-v128-type-i32-expected.txt`, you can write:
```
;;; ERROR: 1
;;; FLAGS: --enable-simd
(module (func v128.const ...
```

lizhengxing commented:
@binji 
Thanks! 
How about test/parse/expr/const.txt?  extract the v128.const to another dedicated test file or use "FLAGS: --enable-simd" in test/parse/expr/const.txt?  

@arunetm FYI

binji commented:
> How about test/parse/expr/const.txt? extract the v128.const to another dedicated test file or use "FLAGS: --enable-simd" in test/parse/expr/const.txt?

It's OK to just add the `--enable-simd` flag to that file.

lizhengxing commented:
@binji 
Thanks!

I updated the test files, please review it again once you are available .

@arunetm FYI!
